File: Algorithms/C++/2-Add_Two_Numbers.cpp
// Runtime: 20 ms, faster than 91.68% of C++ online submissions for Add Two Numbers.
// Memory Usage: 10.2 MB, less than 76.87% of C++ online submissions for Add Two Numbers.
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode res(NULL), *p = &res;
        int flag = 0;
        while (l1 || l2 || flag) {
            int add = (l1 ? l1->val : 0) + (l2 ? l2->val : 0) + flag;
            flag = add >= 10;
            p->next = new ListNode(add % 10);
            p = p->next;
            l1 = l1 ? l1->next : l1;
            l2 = l2 ? l2->next : l2;
        }
        return res.next;
    }
};

File: Algorithms/Java/2-Add_Two_Numbers.java
// Runtime: 2 ms, faster than 77.52% of Java online submissions for Add Two Numbers.
// Memory Usage: 39.2 MB, less than 82.91% of Java online submissions for Add Two Numbers.
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode res = new ListNode();
        ListNode tmp = res;
        boolean flag = false;
        while (l1 != null || l2 != null || flag) {
            int add = (l1 != null ? l1.val : 0) + (l2 != null ? l2.val : 0) + (flag ? 1 : 0);
            flag = add >= 10;
            tmp.next = new ListNode(add % 10);
            tmp = tmp.next;
            l1 = l1 != null ? l1.next : null;
            l2 = l2 != null ? l2.next : null;
        }
        return res.next;
    }
}
